Classes/DDFileLogger.m

```diff
@@ -1531,3 +1540,114 @@ BOOL doesAppRunInBackground() {
 }
 
 #endif
+
+static NSUInteger kMaximumBytesCountInBuffer = (1 << 10) * (1 << 10); // 1 MB.
+static NSUInteger kDefaultBytesCountInBuffer = (1 << 10);
+@interface DDFileLoggerWithBuffer : DDFileLogger
+@property (assign, nonatomic, readwrite) NSUInteger maximumBytesCountInBuffer;
+@end
+
+@interface DDFileLoggerWithBuffer() {
+    NSOutputStream *_bufferStream;
+    NSUInteger _bufferSize;
+}
+@end
+
+@interface DDFileLoggerWithBuffer (StreamManipulation)
+- (void)flushBuffer;
+- (void)dumpBufferToDisk;
+- (void)appendToBufferData:(NSData *)data;
+@end
+
+@implementation DDFileLoggerWithBuffer (StreamManipulation)
+
+- (void)flushBuffer {
+    // do something.
+    [_bufferStream close];
+    _bufferStream = nil;
+    _bufferSize = 0;
+}
+
+- (void)dumpBufferToDisk {
+    // do something.
+    __auto_type data = (NSData *)[_bufferStream propertyForKey:NSStreamDataWrittenToMemoryStreamKey];
+    [super logData:data];
+    [self flushBuffer];
+}
+
+- (void)appendToBufferData:(NSData *)data {
+    __auto_type length = data.length;
+    if (data.length != 0) {
+        if (_bufferStream == nil) {
+            _bufferStream = [[NSOutputStream alloc] initToMemory];
+            [_bufferStream open];
+            _bufferSize = 0;
+        }
+        const uint8_t *appendedData = calloc(length, sizeof(uint8_t));
+        [data getBytes:(void *)appendedData length:length];
+        if (appendedData != NULL) {
+            [_bufferStream write:appendedData maxLength:length];
+        }
+        if (appendedData != NULL) {
+            free((void *)appendedData);
+        }
+        _bufferSize += _bufferSize + length;
+    }
+}
+
+@end
+
+@implementation DDFileLoggerWithBuffer
+
+@synthesize maximumBytesCountInBuffer = _maximumBytesCountInBuffer;
+
+- (instancetype)init {
+    if (self = [super init]) {
+        self.maximumBytesCountInBuffer = kDefaultBytesCountInBuffer;
+    }
+    return self;
+}
+
+#pragma mark - Subclass
+- (void)logData:(NSData *)data {
+    [self putDataIntoBufferOrToDisk:data];
+}
+
+#pragma mark - Initialization
+- (void)cleanup {
+    [self dumpBufferToDisk];
+    [super cleanup];
+}
+
+@end
+
+@interface DDFileLoggerWithBuffer (Logging)
```

Comment:
If we safe us the extra `@interface` and just use `#pragma mark - Logging`, the build won't fail because we use `putDataIntoBufferOrToDisk:` in `logData:` before it's declared 😉

Reply:
@ffried 
`- (void)logData` is not only one place for logging data.
`- (void)cleanup` method which is called in dealloc, also should dump data to disk if necessary.

```objective-c
- (void)cleanup {
    [self dumpBufferToDisk];
    [super cleanup];
}
```

will be expanded to

```
- (void)cleanup {
    __auto_type data = (NSData *)[_bufferStream propertyForKey:NSStreamDataWrittenToMemoryStreamKey];
    [super logData:data];
    [self flushBuffer];
    [super cleanup];
}
```

About locking.
Is it necessary for file logger?
every logger, as I know, runs inside serial queue, no?

Consider `lt_log:`

```
    if (_numProcessors > 1) {
        // Execute each logger concurrently, each within its own queue.
        // All blocks are added to same group.
        // After each block has been queued, wait on group.
        //
        // The waiting ensures that a slow logger doesn't end up with a large queue of pending log messages.
        // This would defeat the purpose of the efforts we made earlier to restrict the max queue size.

        for (DDLoggerNode *loggerNode in self._loggers) {
            // skip the loggers that shouldn't write this message based on the log level

            if (!(logMessage->_flag & loggerNode->_level)) {
                continue;
            }
            
            dispatch_group_async(_loggingGroup, loggerNode->_loggerQueue, ^{ @autoreleasepool {
                [loggerNode->_logger logMessage:logMessage];
            } });
        }
        
        dispatch_group_wait(_loggingGroup, DISPATCH_TIME_FOREVER);
    } else {
        // Execute each logger serialy, each within its own queue.
        
        for (DDLoggerNode *loggerNode in self._loggers) {
            // skip the loggers that shouldn't write this message based on the log level

            if (!(logMessage->_flag & loggerNode->_level)) {
                continue;
            }
            
            dispatch_sync(loggerNode->_loggerQueue, ^{ @autoreleasepool {
                [loggerNode->_logger logMessage:logMessage];
            } });
        }
    }

```

There is serial logging ( with group or with sync ) in both branches.
However, yes, If logger can be added to `loggers` array twice or same loggers could write to same file, some problems could occur.

Reply:
I see. What if we add a method to `DDFileLogger` called `- (void)writeToFile:(NSData *)data` wich does what `- (void)logData:(NSData *)data` currently does? The latter would then just call the former, but the subclass could then simply call `[self writeToFile:data];` instead of "misusing" `super` for this.

About locking: You might be right. I just recalled, that we recently added locking around some properties (e.g. logFileHandle) in DDFileLogger.